### PR TITLE
Feature: Python 3.10 support

### LINKED ISSUE #51: Refactor `loop` parameter usage
The `loop` parameter of most (if not all) `asyncio` function is deprecated and is scheduled to be removed in Python 3.10. We need to carefully check the occurences of the loop parameter in MetricQ and refactor appropiately.

### PR DESCRIPTION
**Pull request by bmario**
This makes MetricQ ready for python 3.10

- Most notably, it refactors the `loop` parameter out. This part needs careful checking. (Fixes #51)
- Adds python 3.10 to test matrix in github actions
- Drops support for all Python versions <3.10

### CODE REVIEW

**Comment by bmario** [comment]
Note: This is a breaking change because it removes the `event_loop` setter.

- [ ] Update migration documentation to 4.0

**Comment by bmario** [comment]
I just noticed that the `Agent` isn't part of the public API, hence, the `event_loop` shouldn't as well.
So... is this even a breaking change then? :thinking:

**Comment by kinnarr** [comment]
Since we pass the `**kwargs` to the parent class each time, it is actually public API. Even some of our examples do it this way (https://github.com/metricq/metricq-python/blob/master/examples/metricq_client.py)

File: metricq/agent.py [hunk 1]
```diff
@@ -142,21 +142,10 @@ def derive_address(self, address: str) -> str:
...
     def event_loop(self) -> asyncio.AbstractEventLoop:
-        # TODO maybe we should instead **TRY** to get the loop of the current context
         if self._event_loop is None:
-            logger.debug("Getting current or new")
-            try:
-                self._event_loop = asyncio.get_event_loop()
-            except RuntimeError:
-                # Until python 3.6, it may raise instead of creating a new one
-                self._event_loop = asyncio.new_event_loop()
+            self._event_loop = asyncio.get_event_loop()
```
Comment (tilsche) [inline metricq/agent.py:146]:
```
def _event_loop(self) -> asyncio.AbstractEventLoop:
    loop = asyncio.get_running_loop()
    if self._loop is None:
        with _global_thread_lock:
            if self._loop is None:
               self._loop = loop
    if loop is not self._loop:
        raise RuntimeError(f'{self!r} is bound to a different event loop')
    return loop
```

+ a module level `_global_thread_lock = threading.Lock()` (that's how asyncio does it :shrug:)
+ `_loop = None` initialization.

And then *always* use this when we need a loop internally.

File: metricq/agent.py [hunk 2]
```diff
@@ -382,7 +370,7 @@ async def rpc(
         request_future = None
 
         if response_callback is None:
-            request_future = self.event_loop.create_future()
+            request_future = asyncio.Future()
```
Comment (tilsche) [inline metricq/agent.py:373]:
```suggestion
            request_future = self._event_loop.create_future()
```

File: metricq/connection_watchdog.py [hunk 3]
```diff
@@ -83,7 +82,7 @@ async def watchdog() -> None:
                 logger.debug("Cancelled {} watchdog", self.connection_name)
                 raise
 
-        self._watchdog_task = loop.create_task(watchdog())
+        self._watchdog_task = asyncio.get_event_loop().create_task(watchdog())
```
Comment (tilsche) [inline metricq/connection_watchdog.py:85]:
```suggestion
        self._watchdog_task = self.event_loop.create_task(watchdog())
```

File: metricq/agent.py [hunk 4]
```diff
@@ -430,7 +418,7 @@ def cleanup() -> None:
                 cleanup()
                 raise te
         elif timeout:
-            self.event_loop.call_later(timeout, cleanup)
+            asyncio.get_running_loop().call_later(timeout, cleanup)
```
Comment (tilsche) [inline metricq/agent.py:421]:
```suggestion
            self._event_loop.call_later(timeout, cleanup)
```

File: tests/test_history_client.py [hunk 5]
```diff
@@ -56,7 +56,7 @@ async def test_history_aggregate(
     history_client: HistoryClient, mocker: MockerFixture
 ) -> None:
     TIME = Timestamp(0)
-    AGGREGATE = create_autospec(history_pb2.HistoryResponse.Aggregate, set_spec=True)
+    AGGREGATE = create_autospec(history_pb2.HistoryResponse.Aggregate, setspec=True)
```
Comment (tilsche) [inline tests/test_history_client.py:59]:
```suggestion
    AGGREGATE = create_autospec(history_pb2.HistoryResponse.Aggregate, spec_set=True)
```

Why does it even work!?!?!

**Review by tilsche — CHANGES_REQUESTED** [review]

File: tests/test_history_client.py (hunk 5 above)
Comment (bmario) [reply tests/test_history_client.py:59]: If you use `set_spec`, you'll get raised an exception with:
`RuntimeError: 'set_spec' might be a typo; use unsafe=True if this is intended`

As that disappeared with removing the underscore, I assumed it's fixed. But don't ask me why the test passes with and without the parameter :(

File: metricq/agent.py (hunk 4 above)
Comment (bmario) [reply metricq/agent.py:428]: I don't have self.event_loop in that class.

**Comment by bmario** [comment]
Rebased onto master and tested with the watchport source, which seems to work now. @tilsche Any more comments?

**Review by tilsche — APPROVED** [review]

**Pull request merged**